<a href="https://colab.research.google.com/github/zengmmm00/DASC_PRE_PYTHON/blob/main/01a_VariableOperatorsIO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> _Self-learning material_  
> **Python workshop - 1a. Variable, operators and basic I/O** 

# Values and Variables

- To maintain program states, we associate values to names.
- These names are called variables.
- In Python, we use the **assignment operator** `=` to assign a value to a variable, for example:

```python
a = 10
```

- In many programming language, variable must be **declared** before it could be used.
  - In Python, assigning a value to a name automatically declare the variable.
- The **assignment operator** `=` will always assign the **value** on the right to the **variable** on the left.
- In the above example, variable `a` is assigned a value of 10.

## Using variables

If we assign a new value to a variable, the variable will be overwritten. For example:

```python
a = 10
a = 20
```

In the above example, variable `a` is assigned a value of 20.

## Types of values

A variable can be used to hold different types of values.
Typical variable type in Python are **boolean**, **integer**, **floating-point**, and **string**.

```python
# boolean values
a = True
b = False
# integer/floating-point value
c = 3
d = 1.5
# string (same for single quote or double quote)
e = "hello"
f = 'world'
```

# Basic output

## Printing multiple values

It is common to print a combination of values and variables. Consider this program for example:

In [ ]:
a = 1
b = 2
c = 3
print(a)
print('+')
print(b)
print('=')
print(c)

1
+
2
=
3


The output is not desirable (try it!). Sometimes, we pefer printing values on the same line.

We can print a list of values, separated by **commas** when we use `print()`, a space will be added between the values:

In [ ]:
a = 1
b = 2
c = 3
print(a, '+', b, '=', c)

1 + 2 = 3


## Separator option: `sep`

We can specify the **`sep`** option in `print()` to specify the separator to be used when printing multiple values.

In [ ]:
print(1, 2, 3)
print(1, 2, 3, sep=',')

1 2 3
1,2,3


## Ending option: `end`

Another option for `print()` is **`end`**, which control how to end the printing.
By default a **new line** is inserted.

In [ ]:
print(1, end=' + ')
print(2, end=' = ')
print(3)

1 + 2 = 3


Sometimes it will be usful to end with an empty string. For example, when we want to output a single line with multiple print statements:

In [ ]:
print(1, end='')
print(2, end='')
print(3)

123


# Arithmetic operators

For integer and floating point numbers, we can use the four **arithmetic operators**: `+`, `-`, `*` (multiply), and `/` (divide). For example:

In [ ]:
print(1 + 2)

3


In [ ]:
print(10 - 1.5)

8.5


In [ ]:
print(2 * 6)

12


In [ ]:
print(9 / 6)

1.5


## Order of execution (operator precedence)

When there are multiple operators in the same statement, the order of execution follows a certain rules. For the basic arithmetic operators, it follows the basic mathematics rules (multiplication and division first), and is processed left-to-right.

In [ ]:
print(1 + 2 * (3 - 4) / 5)

0.6


If unsure, always add brackets to specify the order of execution.

Reference: <https://docs.python.org/3/reference/expressions.html#operator-precedence>

## Floating point division and floor division

Operator `/` always results in a **floating point number**, even when both operands are integers.

In [ ]:
a = 100 
b = 10
print(a / b)

10.0


If **integer division** is needed, we can use the floor division operator (`//`) instead.
This operator will perform division and return the floor of the result.
For example:

In [ ]:
print(9 // 6)

1


In [ ]:
print(100 // 10)

10


## Power operator

The **power operator** `**` calculates and returns the value of a base raised to a specific power. 

Note: the `^` operator in Python is another operator, which will be discussed later (optional).

In [ ]:
print(10 ** 2)

100


In [ ]:
print(2 ** 10)

1024


## Modulo operators

The **modulo** operator `%` calculates and returns the **remainder** of dividing first operand by the second operand.

The two operators, `//` and `%` can be used to find the quotient and remainer of a division operation.

If, `q = a // b` and `r = a % b`, then `a = b * q + r`.

**This operator is extremely important in computer science.** (why?)

In [ ]:
print(100 % 3)

1


In [ ]:
print(100 % 7)

2


# Reading input

The `input()` function will always read a string from user.

In [ ]:
a = input('Please input a string:')
print(a)

Hello


The part `'Please input a string:'` is a message to prompt user for input. It can be omitted:

In [ ]:
a = input()
print(a) 

Hello


## Reading integer or floating-point values

As `input()` will always return a string, before we can use an input value in arithmetic calculation, we need to convert it to integer or floating point values. For example:

In [ ]:
a = int(input())
b = float(input())
print(a, '+', b, '=', a + b)

3 + 0.14 = 3.14


# Quiz

- These questions tests your understanding to the topics.
- All quizzes are available on Moodle for self-evaluation purpose.

## Quiz 1a

1. Name the operator `//`.
2. What is the output of the following program? Try to derive the output without running the code.

```python
print('Hello', 'oh', sep=',', end=' ')
print('my', 'world', sep='', end='!')
```

# Exercises

- Exercises help you to check your understanding to the topics.
- There is no specific requirements to these exercises, you are encouraged to explore different approaches in solving the problems.
- Some exercises are available on Moodle (in the form of VPL exercises) for self-evaluation
  - When doing exercises on VPL, you must follow the exact input/output requirement.
  - You must press the "evaluate" button to evaluate your work for it to be graded.

## Exercise 1-1

- Write a program that reads **two floating-point values** and compute their harmonic mean.
- Harmonic mean \\(H\\) of input values \\(a\\), and \\(b\\) can be calculated by the formula \\(\frac{1}{H}=\frac{1}{2}(\frac{1}{a}+\frac{1}{b})\\), or simply \\(H=\frac{2ab}{a+b}\\).
- Sample input/output:

| Input    | Output |
| ----     | ---    |
| 1<br>4     | 1.6    |
| 3<br>7     | 4.2    |
| 3.7<br>4.3 | 3.9775 |

## Exercise 1-2 (VPL available)

- Write a program that convert time period (in seconds) to the long format represented by the pattern `?h ?m ?s`.
- Sample input/output:

| Input   | Output |
| ----    | ---    |
| 100     | 0h 1m 40s |
| 10000   | 2h 46m 40s |
| 1000000 | 277h 46m 40s |

# Optional: Bitwise operators

## Number base

Apart from base 10 numbers, we can define numbers with base 2, 8 and 16 in Python.
- Base 2 number is prefixed by pattern `0b`.
- Base 8 number is prefixed by pattern `0o`.
- Base 16 number is prefixed by pattern `0x`.

In [ ]:
print(0b011010100)

212


In [ ]:
print(0o324)

212


In [ ]:
print(0xD4)

212


## Binary representation

- In a computer, all values are stored as binary numbers.
- So number `212` is internally stored as `0b11010100`.
- These numbers are left-padded with zeros to match with the bit-length. 
Therefore the number `212` is actually stored as `0b00...011010100`.
- Note that in a signed representation, numbers are stored in [2's complement](https://en.wikipedia.org/wiki/Two%27s_complement) representation.

## Bitwise operations

Bitwise operations apply on numbers bit by bit, for example, the AND operation (`&`) on values 12 (`0b01100`) and 10 (`0b01010`) will be:

```text
  00...01100 (12)
& 00...01010 (10)
------------------
  00...01000 ( 8)
```

Only one of the bits above will give a result of 1 as both operands are 1.

## Bitwise Logical Operators

### AND / OR / XOR

Bitwise logical operators includes AND (`&`), OR (`|`), and XOR (`^`), for example:

In [ ]:
a = 0b01100
b = 0b01010
print(a & b, a | b, a ^ b)

8 14 6


### NOT

There is also the negation operator (`~`) which inverts all the bits. 
For example, positive value `01100` (`00...01100`) will becomes `11...10011`.

In [2's complement](https://en.wikipedia.org/wiki/Two%27s_complement) representation. The above value equals `-13`. 

In [ ]:
a = 0b01100
b = 0b01010
print(~a, ~b)

-13 -11


## Shift Operators

Shift operators shift the binary pattern to the left (`<<`) or right (`>>`). For example:

In [ ]:
x = 0b01101
print(x >> 2, x << 1)

3 26


`x >> 2` shifts value `0b00...01101` (13) two positions to the right, therefore the result is `0b00...011` (3). Similarly, `x << 1` shifts the same value one position to the left, so the result is `0b00...011010` (26).

# Optional: Type hint

## Every value has a "type"

We can check the type of a value using `type()`.

In [ ]:
print(type(1))
print(type(3.14))
print(type('Hello'))

<class 'int'>
<class 'float'>
<class 'str'>


## Variable type

When we assign a value to a variable, the type of variable changes accordingly.

In [ ]:
a = 1
print(type(a))
a = 3.14
print(type(a))
a = 'Hello'
print(type(a))

<class 'int'>
<class 'float'>
<class 'str'>


## Type hint

It is always advised to keep the type of a variable unchanged. In many programming languages, you have to declare the type for a variable before you can use it. In Python, it is not necessary but possible. 

In [ ]:
myInt: int = 1
myFloat: float = 3.14
myStr: str = 'Hello'
print(type(myInt), type(myFloat), type(myStr))

<class 'int'> <class 'float'> <class 'str'>


## But it's just a "hint"

Type hint has no effect during code execution, it is a hint for developers and for the IDE to detect possible error in your program. Variable type will still be changed if you assign value of diffeent type to it.

In [ ]:
myInt: int = 1
myInt = 3.14
print(type(myInt))

<class 'float'>
